web/packages/shared/components/ClusterDropdown/ClusterDropdown.tsx

```diff
@@ -0,0 +1,178 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React, { useState, useEffect } from 'react';
+import { useHistory } from 'react-router';
+import { ButtonSecondary, Flex, Menu, MenuItem, Text } from 'design';
+import { ChevronDown } from 'design/Icon';
+import cfg from 'teleport/config';
+import { Cluster } from 'teleport/services/clusters';
+
+import { HoverTooltip } from 'shared/components/ToolTip';
+
+export interface ClusterDropdownProps {
+  clusterLoader: ClusterLoader;
+  clusterId: string;
+  /*
+   * onLoad is an optional prop. If onLoad is not passed, it will use the built in "loadClusters" function
+   */
+  onLoad?: () => Promise<Cluster[]>;
+  /*
+   * onChange is an optional prop. If onChange is not passed, it will use the built in "changeCluster" function
+   */
+  onChange?: (newValue: string) => void;
+  /*
+   * onError is required because this dropdown can be placed on any page, it does not display it's own error
+   * messages. Even if using the internal "loadClusters", we will pass the error back to be consumed by the parent.
+   */
+  onError: (errorMessage: string) => void;
+}
+
+interface ClusterLoader {
+  fetchClusters: () => Promise<Cluster[]>;
+  clusters: Cluster[];
+}
+
+function createOptions(clusters: Cluster[]) {
+  return clusters.map(cluster => ({
+    value: cluster.clusterId,
+    label: cluster.clusterId,
+  }));
+}
+
+export function ClusterDropdown({
+  clusterLoader,
+  clusterId,
+  onChange,
+  onLoad,
+  onError,
+}: ClusterDropdownProps) {
+  const initialClusters = clusterLoader.clusters;
+  const [options, setOptions] = React.useState<Option[]>(
+    createOptions(initialClusters)
+  );
+  const history = useHistory();
+  const [anchorEl, setAnchorEl] = useState(null);
+
+  const selectedOption = {
+    value: clusterId,
+    label: clusterId,
+  };
+
+  function loadClusters() {
+    onError('');
+    try {
+      return clusterLoader.fetchClusters();
+    } catch (err) {
+      onError(err.message);
+    }
+  }
+
+  function changeCluster(clusterId: string) {
+    const newPathName = cfg.getClusterRoute(clusterId);
+
+    const oldPathName = cfg.getClusterRoute(selectedOption.value);
+
+    const newPath = history.location.pathname.replace(oldPathName, newPathName);
+
+    // keep current view just change the clusterId
+    history.push(newPath);
+  }
+
+  function onChangeOption(clusterId: string) {
+    if (onChange) {
+      onChange(clusterId);
+    } else {
+      changeCluster(clusterId);
+    }
+    handleClose();
+  }
+
+  useEffect(() => {
+    const loadFunc = onLoad || loadClusters;
```

Comment:
I would move the `onError('')` on line 78 here so that it also clears out the error before running a custom `onLoad`.

Reply:
I've removed the custom onLoad so I'll skip this but very great idea if we kept it in.